core/src/main/java/org/elasticsearch/Version.java

```diff
@@ -584,6 +591,15 @@ public Version minimumCompatibilityVersion() {
     }
 
     /**
+     * Returns the maximum hypothetical compatible version based on the current
+     * version. Beta and RC releases are always forward compatible with the maximum
+     * hypothetical version.
+     */
+    public Version maximumCompatibilityVersion() {
```

Comment:
I think this is tricky (thought about it some more). We now say that 2.0 is incompatible with 1.x  but maybe in the future we will decide differently. For example, that 3.0 is compatible with 2.x. We will only make this decision when we come close to 3.0 so we can not "bake it" into 2.0 code now. Since we now add a min compatible check both on the node and master side during join - I don't think we need a max check?